man/man1/pmdbg.1
'\" t
'\"macro stdmacro
.\"
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PMDBG 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmdbg\f1 \- report Performance Co-Pilot debug options
.SH SYNOPSIS
\f3pmdbg\f1
[\f3\-l\f1]
.SH DESCRIPTION
The components of the Performance Co-Pilot (PCP) use
a global mechanism to control diagnostic and debug output.
.PP
.B pmdbg
prints out all the available debug options with
the name of the option and some descriptive text.
.PP
The
.B \-l
option does nothing, but is included for backwards compatibility with older
versions of
.BR pmdbg .
.PP
Most applications using the facilities of the PCP support
a
\f3\-D\f1 \f2name\f1[,\f2name\f1 ...]
command-line syntax to enable debug control using the name(s) of
the desired debug options.
.PP
Alternatively the initial value of the debug control flags may be set
to a comma-separated list if option name(s)
using the environment variable
.BR $PCP_DEBUG ,
provided the applications
use
.BR pmGetOptions (3)
to process command line options and arguments.
If both mechanisms are used the effect is additive, so the resultant flags are
those set via
.B $PCP_DEBUG
combined with those set via any
.B \-D
command line options.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.SH COMPATIBILITY
Prior to PCP 7.0 debug options were maintained in a dual state with
an old-style binary encoded integer (plus macro name for programmers), and a new-style
character string name for each debug option.
.PP
In PCP 7.0 support for the the old-style was dropped.
.PP
The table below documents the equivalence for those debugging
options that appeared in both the the old and the new styles.
.TS
box center;
c   s   s | c
l | r | r | l.
Old Style	New Style
_
Symbolic Name	Mask	Value	Name
_
DBG_TRACE_PDU	0x00000001	1	pdu
DBG_TRACE_FETCH	0x00000002	2	fetch
DBG_TRACE_PROFILE	0x00000004	4	profile
DBG_TRACE_VALUE	0x00000008	8	value
DBG_TRACE_CONTEXT	0x00000010	16	context
DBG_TRACE_INDOM	0x00000020	32	indom
DBG_TRACE_PDUBUF	0x00000040	64	pdubuf
DBG_TRACE_LOG	0x00000080	128	log
DBG_TRACE_LOGMETA	0x00000100	256	logmeta
DBG_TRACE_OPTFETCH	0x00000200	512	optfetch
DBG_TRACE_AF	0x00000400	1024	af
DBG_TRACE_APPL0	0x00000800	2048	appl0
DBG_TRACE_APPL1	0x00001000	4096	appl1
DBG_TRACE_APPL2	0x00002000	8192	appl2
DBG_TRACE_PMNS	0x00004000	16384	pmns
DBG_TRACE_LIBPMDA	0x00008000	32768	libpmda
DBG_TRACE_TIMECONTROL	0x00001000	65536	timecontrol
DBG_TRACE_PMC	0x00002000	131072	pmc
DBG_TRACE_DERIVE	0x00004000	262144	derive
DBG_TRACE_LOCK	0x00008000	524288	lock
DBG_TRACE_INTERP	0x00010000	1048576	interp
DBG_TRACE_CONFIG	0x00020000	2097152	config
DBG_TRACE_PMAPI	0x00040000	4194304	pmapi
DBG_TRACE_FAULT	0x00080000	8388608	fault
DBG_TRACE_AUTH	0x00100000	16777216	auth
DBG_TRACE_DISCOVERY	0x00200000	33554432	discovery
DBG_TRACE_ATTR	0x00400000	67108864	attr
DBG_TRACE_HTTP	0x00800000	134217728	http
DBG_TRACE_DESPERATE	0x04000000	1073741824	desperate
.TE
.SH SEE ALSO
.BR PCPIntro (1),
.BR PMAPI (3),
.BR pmClearDebug (3),
.BR pmGetDebug (3),
.BR pmGetOptions (3),
.BR pmSetDebug (3),
.BR pcp.conf (5)
and
.BR pcp.env (5).


CHANGELOG
pcp-7.0.2 (15 October 2025)
    - Currently work-in-progress.

pcp-7.0.1 (21 September 2025)
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: new per-process metrics agent for macOS
	pmdagfs2: Add support for glockfd debugfs metric collection
	pmdalinux: add NFSv4 connected clients and open files metrics
	pmdalinux: additional /proc/net/netstat MPTcp metrics
	pmdabpftrace: Adjust event loop code to work with Python 3.14
    - Client tools and utilities:
	guidellm2pcp: convert GuideLLM benchmark results to PCP archives
	pcp-mpstat: use vuser/vnice metrics and unify CPU utilization values
    - Server-side utilities and log management scripts:
	pmproxy, libpcp_web: ensure correct path used for conflict resolution
	pmlogger, pmie scripts: improvements for BSD date(1) differences
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix accounting within __pmHashCtl after walk-based-delete
	libpcp: fixes for >2GB archive files on 32-bit platforms
	python api: fix datetime-based timestamps in import API
	python api: improve exception handler in import API
    - Build, infrastructure, containers and packaging updates:
	build: automate macOS release and build process on github
	build: add aarch64 archive-analysis container builds on quay+ghcr
	build: debian conditional Build-Depends: for systemd pkgconfig file
	build: packaging list updates for RHEL 10 builds of PCP
	build: add pmproxy to the rpm set needing LOCAL_ONLY port setup
	build: ensure pcp-zeroconf rpm install script succeeds
	build: add zstd package as a hard rpm package dependency
	containers: archive-analysis container exports optional REST API
	containers: switch archive-analysis to using pcp-import-pmseries
	containers: add PCP_DOMAIN_AGENTS to pcp for auto-enabling PMDAs
	containers: make most/all PMDAs available in the pcp container
	containers: update the pmproxy template config in pcp container
    - Documentation:
	docs: fix openapi.yaml issues (pcp.readthedocs.io/en/latest/api)
	docs: add a containers note about alternate registry at ghcr.io
	docs: add a CLAUDE.md file for devs using claude code with PCP

pcp-7.0.0 (1 September 2025)
    - PMDA additions, enhancements and bug fixes:
	pmdaopentelemetry: new exporter for OpenTelemetry metrics
	pmdarocestat: new RDMA over Converged Ethernet (RoCE) metrics
	pmdahdb: new agent exporting SAP HANA metrics
	pmdalinux: additional TCP networking metrics
	pmdalinux: additional kernel memory and hugetlb metrics
	pmdalinux: support Amazon Linux within kernel.uname.distro
	pmdaopenmetrics: improvements to dynamic metric name handling
	pmdaopenmetrics: allow configuration files with epoch times
	pmdapodman: fix container name/command handling in new podman
	pmdahacluster: additional pacemaker node metrics
	pmdasmart: add smart.wwid.* metrics with drive UUID for indom
	pmdammv: support empty instance domains and dynamic map resets
	pmdammv: export instance domain help text when requested
    - Server-side utilities and log management scripts:
	pmlogger, pmproxy: support for a remote archive push webhook;
	  providing central logging without server-side configuration
	pmlogconf: updates to ensure complete sysstat metric coverage
	pmproxy_daily: new daily script for pmlogger-pushed archives
	pmproxy: support both OpenMetrics and OpenTelemetry /metrics
	pmproxy: configurable HTTP headers (e.g. X-Grafana-Device-ID)
	pmlogger_daily: default compression is a dynamic zst/xz choice
	rc_pcp: legacy init.d pmcd+pmlogger wrapper script removed
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp.so.4: default PMAPI supports nanosecond precision times
	libpcp.so.3: provides backwards compatible libpcp interfaces
	pmapi.h: backwards compatible APIs via -DPMAPI_VERSION=2|4
	libpcp: flag for pmproxy to open archives at the last volume
	libpcp: fix lurking __pmOptFetchAdd bug affecting pmlogger
	libpcp_import: protect against a second call to pmiEnd
	python api: fix (de)reference of a tuple in refresh callback
	python api: removed python version 2 support, 3.6 is minimum
    - Client tools and utilities:
	pcp-rocestat: client to report NVIDIA RoCE metrics
	pcp-dstat: fix the --time-adv option with archives
	pcp-htop: sync-up with latest htop.dev code fixes
	pcp-htop: improve pmcd disconnect/restart handling
	pmrep: fix iostat merge percentage formulas
	pmrep: allow for not preserving discrete metric values
	pmrep: add sar-B and sar-r columns, sync with recent sysstat
	pmjson: produce colorful output to the console by default
	pmlogpush: new tool for pushing existing archives to pmproxy
	pcp2opentelemetry: new tool exporting to OpenTelemetry format
	pmseries_import: new tool for archive analysis container
	archive analysis container: removed rebuild of PCP sources
	archive analysis container: refactored to use real packages
	archive analysis container: set Grafana time window from archive
    - Build, infrastructure and packaging updates:
	build: MacOS builds and packaging, welcome back!
	build: removed Qt4 support from code base, 5.6 is minimum
	build: various improvements for reproducible builds
	build: archive-analysis container reworked to use packages
	build: fixes to better support parallel building
	build: fix multilib installs for pcp-libs-devel packages
	build: provide libvirt PMDA on any el7 if py3 modules exist
	build: clean up debian missing Build-Depends: packages
	build: allow -DNDEBUG to be used consistently across distro builds
	build: allow containers of any architecture to be built
	selinux: update policy for AMD and NVIDIA gpu pmdas
	selinux: additional policy needed for pcp_pmie_t using ps
	selinux: add permissions allowing proc_psi_t access
	tmpfiles.d: additions to support better ostree integration
	pcp-zeroconf: use systemd presets to drive service start
    - Documentation:
	man pages: improved function prototypes in SYNOPSIS section
	man pages: updates to PCPCompat(1) describing major changes


src/pmdbg/pmdbg.c
/*
 * Copyright (c) 2013-2014,2016-2017 Red Hat.
 * Copyright (c) 1995,2002-2003 Silicon Graphics, Inc.  All Rights Reserved.
 * 
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 */

/*
 * pmdbg - help for PCP debug fields
 */

#include <ctype.h>
#include "pmapi.h"
#include "../libpcp/src/pmdbg.h"

static char	*fmt = "%-14.14s  %s";

static pmLongOptions longopts[] = {
    PMAPI_OPTIONS_HEADER("General options"),
    { "list", 0, 'l', 0, "display names and text for all PCP debug options [default action]" },
    PMOPT_HELP,
    PMAPI_OPTIONS_END
};

static pmOptions opts = {
    .short_options = "l?",
    .long_options = longopts,
    .short_usage = "[-l]",
};

int
main(int argc, char **argv)
{
    int		i;
    int		c;

    while ((c = pmgetopt_r(argc, argv, &opts)) != EOF) {
	switch (c) {

	case 'l':	/* list all debug options [same as default] */
	    break;

	case '?':
	default:
	    opts.errors++;
	    break;
	}
    }

    if (opts.errors) {
	pmUsageMessage(&opts);
	exit(1);
    }

    printf("Performance Co-Pilot Debug Options\n");
    printf(fmt, "Option", "Meaning");
    putchar('\n');
    for (i = 0; i < num_debug; i++) {
	printf(fmt, debug_map[i].name, debug_map[i].text);
	putchar('\n');
    }

    return 0;
}
